lib/rules/no-unused-state.js

```diff
@@ -0,0 +1,338 @@
+/**
+ * @fileoverview  Attempts to discover all state fields in a React component and
+ * warn if any of them are never read.
+ *
+ * State field definitions are collected from `this.state = {}` assignments in
+ * the constructor, objects passed to `this.setState()`, and `state = {}` class
+ * property assignments.
+ */
+
+'use strict';
+
+var Components = require('../util/Components');
+
+// Descend through all wrapping TypeCastExpressions and return the expression
+// that was cast.
+function uncast(node) {
+  while (node.type === 'TypeCastExpression') {
+    node = node.expression;
+  }
+  return node;
+}
+
+// Return the name of an identifier or the string value of a literal. Useful
+// anywhere that a literal may be used as a key (e.g., member expressions,
+// method definitions, ObjectExpression property keys).
+function getName(node) {
+  node = uncast(node);
+  if (node.type === 'Identifier') {
+    return node.name;
+  } else if (node.type === 'Literal') {
+    return String(node.value);
+  }
+  return null;
+}
+
+function isThisExpression(node) {
+  return uncast(node).type === 'ThisExpression';
+}
+
+function getInitialClassInfo() {
+  return {
+    // Set of nodes where state fields were defined.
+    stateFields: [],
+
+    // Set of names of state fields that we've seen used.
+    usedStateFields: [],
+
+    // Names of local variables that may be pointing to this.state. To
+    // track this properly, we would need to keep track of all locals,
+    // shadowing, assignments, etc. To keep things simple, we only
+    // maintain one set of aliases per method and accept that it will
+    // produce some false negatives.
+    aliases: null
+  };
+}
+
+module.exports = {
+  meta: {
+    docs: {
+      description: 'Prevent definition of unused state fields',
+      category: 'Best Practices',
+      recommended: false
+    },
+    schema: []
+  },
+
+  create: Components.detect(function(context, components, utils) {
+    // Non-null when we are inside a React component ClassDeclaration and we have
+    // not yet encountered any use of this.state which we have chosen not to
+    // analyze. If we encounter any such usage (like this.state being spread as
+    // JSX attributes), then this is again set to null.
+    var classInfo = null;
+
+    // Returns true if the given node is possibly a reference to `this.state`.
+    function isStateReference(node) {
+      node = uncast(node);
+
+      var isDirectStateReference =
+        node.type === 'MemberExpression' &&
+        isThisExpression(node.object) &&
+        node.property.name === 'state';
```

Comment:
Yep, every `MemberExpression` should have a `property` AFAIK. It may not have a `name` (if it's computed), but it won't crash.

If we used `getName(node.property) === 'state'` instead we'd also handle the computed case (so we'd detect uses like `this['state']`). Looks like I forgot to do that here :)